File: app/src/main/java/com/jmalinen/blackjack/engine/PayoutCalculator.kt
package com.jmalinen.blackjack.engine

import com.jmalinen.blackjack.model.CasinoRules
import com.jmalinen.blackjack.model.Hand
import com.jmalinen.blackjack.model.HandResult

object PayoutCalculator {

    data class RoundResult(
        val handResults: Map<Int, HandResult>,
        val totalPayout: Int,
        val insurancePayout: Int
    )

    fun calculateResults(
        playerHands: List<Hand>,
        dealerHand: Hand,
        insuranceBet: Int,
        rules: CasinoRules
    ): RoundResult {
        var totalPayout = 0
        var insurancePayout = 0
        val results = mutableMapOf<Int, HandResult>()

        if (insuranceBet > 0 && dealerHand.isBlackjack) {
            insurancePayout = insuranceBet * 3
            totalPayout += insurancePayout
        }

        for ((index, hand) in playerHands.withIndex()) {
            val result: HandResult
            val payout: Int

            when {
                hand.isSurrendered -> {
                    result = HandResult.SURRENDER
                    payout = hand.bet / 2
                }
                hand.isBusted -> {
                    result = HandResult.BUST
                    payout = 0
                }
                hand.isBlackjack && !dealerHand.isBlackjack -> {
                    result = HandResult.BLACKJACK
                    payout = hand.bet + (hand.bet * rules.blackjackPayout.multiplier).toInt()
                }
                hand.isBlackjack && dealerHand.isBlackjack -> {
                    result = HandResult.PUSH
                    payout = hand.bet
                }
                !hand.isBlackjack && dealerHand.isBlackjack -> {
                    result = HandResult.LOSE
                    payout = 0
                }
                dealerHand.isBusted -> {
                    result = HandResult.WIN
                    payout = hand.bet * 2
                }
                hand.score > dealerHand.score -> {
                    result = HandResult.WIN
                    payout = hand.bet * 2
                }
                hand.score == dealerHand.score -> {
                    result = HandResult.PUSH
                    payout = hand.bet
                }
                else -> {
                    result = HandResult.LOSE
                    payout = 0
                }
            }
            results[index] = result
            totalPayout += payout
        }

        return RoundResult(results, totalPayout, insurancePayout)
    }
}

File: ios/Blackjack/Blackjack/Engine/PayoutCalculator.swift
enum PayoutCalculator {

    struct RoundResult: Equatable {
        let handResults: [Int: HandResult]
        let totalPayout: Int
        let insurancePayout: Int
    }

    static func calculateResults(
        playerHands: [Hand],
        dealerHand: Hand,
        insuranceBet: Int,
        rules: CasinoRules
    ) -> RoundResult {
        var totalPayout = 0
        var insurancePayout = 0
        var results = [Int: HandResult]()

        if insuranceBet > 0 && dealerHand.isBlackjack {
            insurancePayout = insuranceBet * 3
            totalPayout += insurancePayout
        }

        for (index, hand) in playerHands.enumerated() {
            let result: HandResult
            let payout: Int

            let isThreeSevens = rules.threeSevensPays3to1 &&
                hand.cards.count == 3 &&
                hand.cards.allSatisfy { $0.rank == .seven }

            if hand.isSurrendered {
                result = .surrender
                payout = hand.bet / 2
            } else if hand.isBusted {
                result = .bust
                payout = 0
            } else if isThreeSevens {
                result = .threeSevens
                payout = hand.bet * 4
            } else if hand.isBlackjack && !dealerHand.isBlackjack {
                result = .blackjack
                payout = hand.bet + Int(Float(hand.bet) * rules.blackjackPayout.multiplier)
            } else if hand.isBlackjack && dealerHand.isBlackjack {
                result = .push
                payout = hand.bet
            } else if !hand.isBlackjack && dealerHand.isBlackjack {
                result = .lose
                payout = 0
            } else if dealerHand.isBusted {
                result = .win
                payout = hand.bet * 2
            } else if hand.score > dealerHand.score {
                result = .win
                payout = hand.bet * 2
            } else if hand.score == dealerHand.score {
                result = .push
                payout = hand.bet
            } else {
                result = .lose
                payout = 0
            }

            results[index] = result
            totalPayout += payout
        }

        return RoundResult(handResults: results, totalPayout: totalPayout, insurancePayout: insurancePayout)
    }
}
